http_server/parse_start.py

```diff
@@ -0,0 +1,138 @@
+from amaranth import Module
+from amaranth.lib.wiring import In, Out, Component, connect, flipped
+from amaranth.lib import stream
+
+from string_contains_match import StringContainsMatch
+from stream_demux import StreamDemux
+import stream_utils
+
+class ParseStart(Component):
+    """
+    Parser for the start-line of an HTTP request.
+
+    Parameters
+    ----------
+    paths: list[str]
+           Valid paths to match
+
+    Attributes
+    ----------
+    input:    Stream(8), in
+              Data stream to match
+    reset:    Signal(1), in
+              Reset and await new input
+    done:     Signal(1), out
+              Indicates that the "\r\n" end-of-line sequence was seen.
+    method:   list(Signal(1)), out
+              Bitfield of matched methods. The 0th field indicates no match.
+              METHOD_* constants can be used for decode.
+    path:     list(Signal(1)), out
+              Bitfield of matched paths. The 0th field indicates no match.
+              Other matches are in the order from the paths parameter.
+    protocol: list(Siganl(1)), out
+              Bitfield of matched protocol. The 0th field indicates no match.
+              PROTOCOL_* constants can be used for decode.
+    """
+
+    METHOD_NO_MATCH = 0
+    METHOD_GET = 1
+    METHOD_POST = 2
+
+    PROTOCOL_NO_MATCH = 0
+    PROTOCOL_HTTP1_0 = 1
+
+    def __init__(self, paths):
+        super().__init__({
+            "input": In(stream.Signature(8)),
+            "reset": In(1),
+            "done": Out(1),
+            "method": Out(3),
+            "path": Out(len(paths)+1),
+            "protocol": Out(2),
+        })
+        self._paths = paths
+
+    def elaborate(self, _platform):
+        m = Module()
+
+        resets = []
+
+        method_stream = stream.Signature(8).create()
+        get_matcher = m.submodules.get_matcher = StringContainsMatch("GET")
+        resets.append(get_matcher.reset)
+        m.d.comb += self.method[self.METHOD_GET].eq(get_matcher.accepted)
+
+        post_matcher = m.submodules.post_matcher = StringContainsMatch("POST")

```

Comment:
Not necessarily this PR, but a TODO:

We might want to use an exact match for this -- the request line is `^(GET|POST|TRACE|...)`, not `.*POST`. "Consume until all request matchers have accepted or rejected"?

(See https://en.wikipedia.org/wiki/HTTP_request_smuggling)